### PR TITLE
Ninja: Update deps, use HikariCP instead of c3p0 db pool

### PR DESCRIPTION
**Pull request by k-r-g**
HikariCP is significantly faster than c3p0 in published benchmarks and my crude testing.

I also updated dependencies to the latest versions and fixed a bug where the updates test wasn't capping the random number.

@raphaelbauer Is this change reasonable?

### CODE REVIEW

**Comment by raphaelbauer** [comment]
Very clever - looks awesome!

What about porting that change to Ninja itself? Should not be a big change, and has a big upside... (I guess only the archetype and the integration test has to be changed).

**Comment by k-r-g** [comment]
@raphaelbauer Haha, updating Ninja is of course totally up to you. What is your preference? Would you like to cancel this PR and create a new one updating Ninja itself? (Note: If you do that, let's make sure we don't lose track of the random number fix in this PR.)

File: frameworks/Java/ninja-standalone/src/main/resources/META-INF/persistence.xml [hunk 1]
```diff
@@ -20,15 +20,16 @@
...
                 <!-- Connection Pooling settings -->
-                <property name="hibernate.connection.provider_class"
-                          value="org.hibernate.service.jdbc.connections.internal.C3P0ConnectionProvider" />
-
-                <property name="hibernate.c3p0.max_size" value="100" />
-                <property name="hibernate.c3p0.min_size" value="0" />
-                <property name="hibernate.c3p0.acquire_increment" value="1" />
-                <property name="hibernate.c3p0.idle_test_period" value="300" />
-                <property name="hibernate.c3p0.max_statements" value="0" />
-                <property name="hibernate.c3p0.timeout" value="100" />
```
Comment (raphaelbauer) [inline frameworks/Java/ninja-standalone/src/main/resources/META-INF/persistence.xml]: Did you run mvn clean install? This should run tests and tests should pass without any issues... But it did not work for me... I had to rewrite the h2 part to:

        <persistence-unit name="h2" transaction-type="RESOURCE_LOCAL">
            <provider>org.hibernate.ejb.HibernatePersistence</provider>
            <properties>
                <property name="javax.persistence.provider" value="org.hibernate.ejb.HibernatePersistence" />
                <property name="hibernate.connection.driver_class" value="org.h2.Driver" />
                <property name="hibernate.dialect" value="org.hibernate.dialect.H2Dialect" />
                
                <property name="hibernate.hbm2ddl.auto" value="create" />
                <property name="hibernate.show_sql" value="false" />
                <property name="hibernate.format_sql" value="false" />
                <property name="hibernate.jdbc.batch_size" value="100" />

                
                <property name="hibernate.connection.provider_class" value="org.hibernate.hikaricp.internal.HikariCPConnectionProvider" />
            </properties>
        </persistence-unit>

File: frameworks/Java/ninja-standalone/src/main/resources/META-INF/persistence.xml [hunk 2]
```diff
@@ -44,15 +45,16 @@
...
-				value="org.hibernate.service.jdbc.connections.internal.C3P0ConnectionProvider" />
-
-			<property name="hibernate.c3p0.max_size" value="100" />
-			<property name="hibernate.c3p0.min_size" value="0" />
-			<property name="hibernate.c3p0.acquire_increment" value="1" />
-			<property name="hibernate.c3p0.idle_test_period" value="300" />
-			<property name="hibernate.c3p0.max_statements" value="0" />
-			<property name="hibernate.c3p0.timeout" value="100" />
+                        <property name="hibernate.connection.provider_class" value="org.hibernate.hikaricp.internal.HikariCPConnectionProvider" />
+                        <property name="hibernate.dialect" value="org.hibernate.dialect.MySQLDialect" />
```
Comment (raphaelbauer) [inline frameworks/Java/ninja-standalone/src/main/resources/META-INF/persistence.xml]: hibernate.dialect is already set some lines above...

File: frameworks/Java/ninja-standalone/src/main/resources/META-INF/persistence.xml [hunk 3]
```diff
@@ -44,15 +45,16 @@
...
+                        <property name="hibernate.cache.use_query_cache" value="false" />
+                        <property name="hibernate.hikari.minimumIdle" value="256" />
+                        <property name="hibernate.hikari.maximumPoolSize" value="256" />
+                        <property name="hibernate.hikari.idleTimeout" value="30000" />
+                        <property name="hibernate.hikari.dataSourceClassName" value="com.mysql.jdbc.jdbc2.optional.MysqlDataSource" />
+                        <property name="hibernate.hikari.dataSource.url" value="jdbc:mysql://localhost:3306/hello_world?jdbcCompliantTruncation=false&amp;elideSetAutoCommits=true&amp;useLocalSessionState=true&amp;cachePrepStmts=true&amp;cacheCallableStmts=true&amp;alwaysSendSetIsolation=false&amp;prepStmtCacheSize=4096&amp;cacheServerConfiguration=true&amp;prepStmtCacheSqlLimit=2048&amp;zeroDateTimeBehavior=convertToNull&amp;traceProtocol=false&amp;useUnbufferedInput=false&amp;useReadAheadInput=false&amp;maintainTimeStats=false&amp;useServerPrepStmts&amp;cacheRSMetadata=true" />
```
Comment (raphaelbauer) [inline frameworks/Java/ninja-standalone/src/main/resources/META-INF/persistence.xml]: url, user and password are usually set in src/main/java/conf/application.conf

The following 4 lines are relevant:
    %prod.ninja.jpa.persistence_unit_name = mysql
    %prod.db.connection.url=jdbc:mysql://localhost:3306/hello_world?jdbcCompliantTruncation=false&amp;elideSetAutoCommits=true&amp;useLocalSessionState=true&amp;cachePrepStmts=true&amp;cacheCallableStmts=true&amp;alwaysSendSetIsolation=false&amp;prepStmtCacheSize=4096&amp;cacheServerConfiguration=true&amp;prepStmtCacheSqlLimit=2048&amp;zeroDateTimeBehavior=convertToNull&amp;traceProtocol=false&amp;useUnbufferedInput=false&amp;useReadAheadInput=false&amp;maintainTimeStats=false&amp;useServerPrepStmts&amp;cacheRSMetadata=true
    %prod.db.connection.username=benchmarkdbuser
    %prod.db.connection.password=benchmarkdbpass


I am not an expert in hibernate's hikari implementation, but I guess you can simply use configuration in application.conf and remove config via persistence.xml...

File: frameworks/Java/ninja-standalone/src/main/resources/META-INF/persistence.xml [hunk 4]
```diff
@@ -44,15 +45,16 @@
...
+                        <property name="hibernate.hikari.dataSource.user" value="benchmarkdbuser" />
+                        <property name="hibernate.hikari.dataSource.password" value="benchmarkdbpass" />
```
Comment (raphaelbauer) [inline frameworks/Java/ninja-standalone/src/main/resources/META-INF/persistence.xml]: Did you try the production configuration? I did not run it with mysql... So I can't comment...

**Comment by raphaelbauer** [comment]
Hey @k-r-g , I added some remarks to persistence.xml. Let me know if that makes sense... Great stuff :+1:

**Comment by k-r-g** [comment]
@raphaelbauer I'm certainly not an expert with HikariCP, but this all looks reasonable to me!

**Comment by raphaelbauer** [comment]
Did you run it against a real mysql db, or was it just a sketch?

**Comment by k-r-g** [comment]
I ran it against a local MySQL instance, and that seems to work fine.

I just tried "mvn clean install" and I get test failures. I had not tried those before.

This is what I see:
Tests in error: 
  multipleQueries(controllers.HelloDbControllerTest): ninja.lifecycle.FailedStartException: java.lang.NoClassDefFoundError: org/w3c/dom/ElementTraversal
  multipleQueries(controllers.HelloDbControllerTest)
  testSingleGet(controllers.HelloDbControllerTest): ninja.lifecycle.FailedStartException: java.lang.NoClassDefFoundError: org/w3c/dom/ElementTraversal
  testSingleGet(controllers.HelloDbControllerTest)
  testUpdates(controllers.HelloDbControllerTest): ninja.lifecycle.FailedStartException: java.lang.NoClassDefFoundError: org/w3c/dom/ElementTraversal
  testUpdates(controllers.HelloDbControllerTest)
  helloPlaintextControllerTest(controllers.HelloPlaintextControllerTest): ninja.lifecycle.FailedStartException: java.lang.NoClassDefFoundError: org/w3c/dom/ElementTraversal
  helloPlaintextControllerTest(controllers.HelloPlaintextControllerTest)
  testHelloJsonController(controllers.HelloJsonControllerTest): ninja.lifecycle.FailedStartException: java.lang.NoClassDefFoundError: org/w3c/dom/ElementTraversal
  testHelloJsonController(controllers.HelloJsonControllerTest)
  testSomeMethod(controllers.HelloFortuneControllerTest): ninja.lifecycle.FailedStartException: java.lang.NoClassDefFoundError: org/w3c/dom/ElementTraversal
  testSomeMethod(controllers.HelloFortuneControllerTest)

Is this a real problem, or do the tests just need to be fixed up?

**Comment by raphaelbauer** [comment]
That's a real problem.

You have to add

    <dependency>
        <groupId>xml-apis</groupId>
        <artifactId>xml-apis</artifactId>
        <version>1.4.01</version>
    </dependency>

to the dependencies....

**Comment by k-r-g** [comment]
@raphaelbauer "mvn clean install" works for me locally now. I had messed up the h2 database connection. Hopefully this will pass in Travis.

File: toolset/setup/linux/client.sh [hunk 5]
```diff
@@ -21,9 +21,9 @@ sudo sh -c "echo '*               -    nofile          65535' >> /etc/security/l
 # wrk
 ##############################
 
-git clone https://github.com/wg/wrk.git
-cd wrk
-git checkout 205a1960c8b8de5f500bb143863ae293456b7add
+curl -sL -o wrk-4.0.1.tar.gz https://github.com/wg/wrk/archive/4.0.1.tar.gz
```
Comment (msmith-techempower) [inline toolset/setup/linux/client.sh]: Why are these changes here?

**Comment by msmith-techempower** [comment]
Whatever, Travis tests passed; LGTM!

**Comment by raphaelbauer** [comment]
Thanks!

**Comment by k-r-g** [comment]
@brettwooldridge Now that Round 12 is out, we have something to compare. Since Round 12 was rushed, however, it's likely some stuff is not quite right and could be improved--probably I messed something up. :)

Below are some quick notes on how HikariCP impacted things. 

These framework tests changed from c3p0 to HikariCP:
* compojure raw
* http-kit raw
* grizzly-jersey
* ninja

Also, these 2 new tests allow comparison of c3p0 and HikariCP:
* undertow-jersey-hikaricp
* undertow-jersey-c3p0


# Test: Single Query
https://www.techempower.com/benchmarks/#section=data-r11&hw=peak&test=db&f=oyzgk-0-27wr28-0-0-cn4
https://www.techempower.com/benchmarks/#section=data-r12&hw=peak&test=db&f=oyzgk-0-27wr28-0-0-cn4

* compojure and http-kit aren't comparable here due to https://github.com/TechEmpower/FrameworkBenchmarks/pull/1896
* grizzly-jersey increased from 53,956 to 113,084
* ninja increased from 28,397 to 34,429
* undertow-jersey increased from 52,196 to 100,219

# Test: Multiple Queries
https://www.techempower.com/benchmarks/#section=data-r11&hw=peak&test=query&f=oyzgk-0-27wr28-0-0-cn4
https://www.techempower.com/benchmarks/#section=data-r12&hw=peak&test=query&f=oyzgk-0-27wr28-0-0-cn4

* compojure-raw increased from 1,390 to 3,329
* http-kit raw didn't exist in R11
* grizzly-jersey increased from 3,376 to 8,791
* ninja **decreased** from 7,905 to 3,948. Weird! I don't know what happened here.
* undertow-jersey increased from 1,500 to 3,556

# Test: Fortunes
https://www.techempower.com/benchmarks/#section=data-r11&hw=peak&test=fortune&f=oyzgk-0-27wr28-0-0-cn4
https://www.techempower.com/benchmarks/#section=data-r12&hw=peak&test=fortune&f=oyzgk-0-27wr28-0-0-cn4

Once again compojure and http-kit aren't comparable here due to https://github.com/TechEmpower/FrameworkBenchmarks/pull/1896

* grizzly-jersey increased from 46,344 to 52,801
* ninja increased from 15,995 to 17,572
* undertow-jersey increased from 26,271 to 29,920

# Test: Data Updates
https://www.techempower.com/benchmarks/#section=data-r11&hw=peak&test=update&f=oyzgk-0-27wr28-0-0-cn4
https://www.techempower.com/benchmarks/#section=data-r12&hw=peak&test=update&f=oyzgk-0-27wr28-0-0-cn4

* compojure increased from 1,115 to 1,300
* http-kit raw didn't exist in R11
* grizzly-jersey doesn't implement this test
* ninja **decreased** from 2,630 to 2,165. Again, I don't know why. Notably, ninja in both rounds had a significant error count.
* undertow-jersey doesn't implement this test

@raphaelbauer Any ideas why Ninja shows errors on the Data Updates test, or why HikariCP seems to have hurt Ninja's performance on Multiple Queries and Data Updates?

**Comment by k-r-g** [comment]
Also, I should note that various other dependencies were changed between Rounds 11 and 12 (e.g., Clojure was updated to 1.8.0) so there are confounding variables.
